mrbean/src/test/java/com/fasterxml/jackson/module/mrbean/TestBridgeMethods.java

```diff
@@ -0,0 +1,302 @@
+import net.bytebuddy.jar.asm.Label;
+import net.bytebuddy.jar.asm.MethodVisitor;
+import net.bytebuddy.jar.asm.TypePath;
+import net.bytebuddy.utility.OpenedClassReader;
+
+import java.io.IOException;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ *
+ */
+public class TestBridgeMethods extends BaseTest {
+
+    public interface Drink {
+        String getType();
+    }
+
+    public interface Coffee extends Drink {
+        String getFlavor();
+    }
+
+    public interface DrinkHolder {
+        Drink getDrink();
+    }
+
+    public interface CoffeeHolder extends DrinkHolder {
+        Coffee getDrink();
+    }
+
+    public interface GenericHolder<T> {
+        T getObject();
+    }
+
+    public interface SpecificCoffeeHolder extends GenericHolder<Coffee> {
+        @JsonProperty("drink")
+        Coffee getObject();
+    }
+
+    public void testSimpleCovariantProperty() throws Exception
+    {
+        ObjectMapper mapper = newMrBeanMapper();
+
+        Class<? extends CoffeeHolder> targetClass = reorderBridgeMethodFirst(CoffeeHolder.class, "getDrink");
+        CoffeeHolder result = mapper.readValue("{\"drink\":{\"type\":\"coffee\",\"flavor\":\"pumpkin spice\"}}", targetClass);
+        assertNotNull(result);
+        assertNotNull(result.getDrink());
+        assertEquals("coffee", result.getDrink().getType());
+        assertEquals("pumpkin spice", result.getDrink().getFlavor());
+    }
+
+    public void testGenericCovariantProperty() throws Exception
+    {
+        ObjectMapper mapper = newMrBeanMapper();
+
+        Class<? extends SpecificCoffeeHolder> targetClass = reorderBridgeMethodFirst(SpecificCoffeeHolder.class, "getObject");
+        SpecificCoffeeHolder result = mapper.readValue("{\"drink\":{\"type\":\"coffee\",\"flavor\":\"pumpkin spice\"}}", targetClass);
+        assertNotNull(result);
+        assertNotNull(result.getObject());
+        assertEquals("coffee", result.getObject().getType());
+        assertEquals("pumpkin spice", result.getObject().getFlavor());
+    }
+
+    /**
+     * Rewrites the specified class so that the bridge version of the specified method appears *before* the non-bridge version.
+     *
+     * This is generally necessary (although maybe not sufficient) for reproducing the issue because:
+     * <ul>
+     *     <li>{@link Class#getDeclaredMethods()} <a href="https://bugs.java.com/bugdatabase/view_bug.do?bug_id=7023180">does not have a predictable order</a>.</li>
+     *     <li>The Java compiler tends to place bridge methods after the actual non-bridge method.</li>
+     *     <li>The order of methods in the class definition typically influences the order seen in reflection.</li>
+     * </ul>
+     */
+    @SuppressWarnings("unchecked")
+    private static <T> Class<T> reorderBridgeMethodFirst(Class<T> clazz, final String methodName) throws IOException {
```

Comment:
I'd love feedback if there is an easier way of doing this.  I'm not very familiar with ASM, but I couldn't find any simpler APIs that let me reorder definitions within the class file.